### ISSUE TITLE
Pulumi stack stuck after a resource provisioning issue

### ISSUE THREAD

**Issue by sylver** [pulumi/pulumi#5377]
An `aws:eks:Cluster` provisioning failed with 

> error: post-step event returned an error: failed to save snapshot: performing HTTP request: Post "https://api.pulumi.com/api/stacks/[REDACTED]/renew_lease": dial tcp 54.69.134.167:443: connect: connection refused

That's a random benign issue from network and/or AWS, a simple retry would work.

The problem is that now the stack is messed up :

> error: the current deployment has 1 resource(s) with pending operations:
>   * urn:pulumi:[REDACTED]$eks:index:Cluster$aws:eks/cluster:Cluster::[REDACTED], interrupted while creating
> 
> These resources are in an unknown state because the Pulumi CLI was interrupted while
> waiting for changes to these resources to complete. You should confirm whether or not the
> operations listed completed successfully by checking the state of the appropriate provider.
> For example, if you are using AWS, you can confirm using the AWS Console.
> 
> Once you have confirmed the status of the interrupted operations, you can repair your stack
> using 'pulumi stack export' to export your stack to a file. For each operation that succeeded,
> remove that operation from the "pending_operations" section of the file. Once this is complete,
> use 'pulumi stack import' to import the repaired stack.
> 
> refusing to proceed

Ok, let's follow what is suggested here : 
- checking the AWS Console, the cluster was properly provisioned
- doing a `pulumi stack export > stack.json`
- "For each operation that succeeded, remove that operation from the `pending_operations` section of the file" => Ok, removed the cluster from `pending_operations`, which is now an empty array in the json file
- doing a `pulumi stack import < stack.json`

Expecting that would fix the stack and be able to pick up the provisioning where we left it off, but now a `pulumi update` gives this error : 

> error: error creating EKS Cluster ([REDACTED]): ResourceInUseException: Cluster already exists with name: [REDACTED]
>     {
>       RespMetadata: {
>         StatusCode: 409,
>         RequestID: "[REDACTED]"
>       },
>       ClusterName: "[REDACTED]",
>       Message_: "Cluster already exists with name: [REDACTED]"
>     }

Tried a `pulumi refresh` but that does not fix the problem, so we're now stuck with a stale stack, and I don't see a way to tell it the cluster was provisioned and move on (the first error suggestion with the stack export does not work as expected, obviously).

the only solution would be to `pulumi destroy` and start fresh, which is not really something we'd like to do if that would happen in production.

**Comment by haiderimrans** [comment]
any update here?

**Comment by floknip** [comment]
> Ok, let's follow what is suggested here :
> 
> * checking the AWS Console, the cluster was properly provisioned
> * doing a `pulumi stack export > stack.json`
> * "For each operation that succeeded, remove that operation from the `pending_operations` section of the file" => Ok, removed the cluster from `pending_operations`, which is now an empty array in the json file
> * doing a `pulumi stack import < stack.json`

In my case ,deleting every entry from the `pending_operations` array inside the `stack.json` did the job. 

Issue has been caused by canceling (cmd+c) the `pulumi up` process in the terminal. 😅

**Comment by guilleijo** [comment]
> > Ok, let's follow what is suggested here :
> > 
> > * checking the AWS Console, the cluster was properly provisioned
> > * doing a `pulumi stack export > stack.json`
> > * "For each operation that succeeded, remove that operation from the `pending_operations` section of the file" => Ok, removed the cluster from `pending_operations`, which is now an empty array in the json file
> > * doing a `pulumi stack import < stack.json`
> 
> In my case, deleting every entry from the `pending_operations` array inside the `stack.json` did the job.
> 
> Issue has been caused by canceling (cmd+c) the `pulumi up` process in the terminal. 😅

Thanks @floknip, this worked for me.

**Comment by snolangps** [comment]
What about if you don't have any `pending_operations` but you're still seeing the `An update is in progress for resource` error? That is the situation I am in...

**Comment by Gwojda** [comment]
Same problem here after a timeout on nodepool replacement.
Everytime i run `pulumi up` its try to fetch and update the nodepool status, which is OK on gcp, but timeout on pulumi after 30mins.
no pending_operations present on stack export.

**Comment by mikhailshilkov** [comment]
There were a few themes and questions in the comments here...

> error: post-step event returned an error: failed to save snapshot: performing HTTP request: Post "[https://api.pulumi.com/api/stacks/[REDACTED]/renew_lease](https://api.pulumi.com/api/stacks/%5BREDACTED%5D/renew_lease)": dial tcp 54.69.134.167:443: connect: connection refused

This is not a resource provisioning error but a network connectivity error. Either Pulumi service was down, or the network failed somewhere in between. There isn't much we can do beyond a few retries (which we already do). This error has nothing to do with AWS.

> urn:pulumi:[REDACTED]$eks:index:Cluster$aws:eks/cluster:Cluster::[REDACTED], interrupted while creating

and a few following comments are focusing on pending operations. `pending_operations` improvements are tracked in #4265.

> Tried a pulumi refresh but that does not fix the problem, so we're now stuck with a stale stack, and I don't see a way to tell it the cluster was provisioned and move on 

This happened because Pulumi doesn't know the unique ID of the created cluster (arn) and therefore no way to track the existing resource to what it needs. And Pulumi state was lost due to the error above. One alternative is to [import](https://www.pulumi.com/docs/guides/adopting/import/) this EKS cluster back to Pulumi state.

> What about if you don't have any pending_operations but you're still seeing the An update is in progress for resource error? That is the situation I am in...

`pulumi cancel` should help here - it would be nice to get a repro for why it didn't. But that's a question that is very different from the original issue - I'd love to discuss it in a separate issue.

> Everytime i run pulumi up its try to fetch and update the nodepool status, which is OK on gcp, but timeout on pulumi after 30mins.

You should be able to solve it with [custom timeouts](https://www.pulumi.com/docs/intro/concepts/resources/options/customtimeouts/).

I'll make a judgment call and decide this issue is overall a duplicate of #4265 with a few side discussions. I'll go ahead and close it as such.
